File: packages/android/src/main/kotlin/gov/civui/components/CivDeceasedPerson.kt
// CivUI — CivDeceasedPerson for Jetpack Compose
// Compound field for collecting information about a person who has died.
// Used on VA burial (21P-530), SSA survivor benefit, and probate forms.
// Plain-language labels — "the person who died", not "decedent".

package gov.civui.components

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import gov.civui.tokens.CivTokens

/** Structured value matching the web `DeceasedPersonValue` interface. */
data class DeceasedPersonValue(
    val first: String = "",
    val middle: String = "",
    val last: String = "",
    val suffix: String = "",
    val dateOfBirth: String = "",
    val dateOfDeath: String = "",
    val relationship: String = "",
)

private val RELATIONSHIP_OPTIONS = listOf(
    "" to "- Select -",
    "spouse" to "Spouse",
    "parent" to "Parent",
    "child" to "Child",
    "sibling" to "Sibling",
    "other" to "Other",
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CivDeceasedPerson(
    value: DeceasedPersonValue,
    onChange: (DeceasedPersonValue) -> Unit,
    modifier: Modifier = Modifier,
    legend: String = "About the person who died",
    name: String = "deceased",
    hint: String = "",
    error: String = "",
    nameError: String = "",
    dateOfBirthError: String = "",
    dateOfDeathError: String = "",
    relationshipError: String = "",
    hideRelationship: Boolean = false,
    required: Boolean = false,
    disabled: Boolean = false,
    readonly: Boolean = false,
) {
    val isDark = isSystemInDarkTheme()
    val errorColor = if (isDark) CivTokens.DarkColors.Error.default_ else CivTokens.Colors.Error.default_
    val hintColor = if (isDark) CivTokens.DarkColors.Base.dark else CivTokens.Colors.Base.dark

    var relationshipExpanded by remember { mutableStateOf(false) }
    val relationshipLabel = RELATIONSHIP_OPTIONS.firstOrNull { it.first == value.relationship }?.second ?: ""

    Column(
        verticalArrangement = Arrangement.spacedBy(CivTokens.Spacing._4),
        modifier = modifier,
    ) {
        if (legend.isNotEmpty()) {
            Text(
                legend + (if (required) " (required)" else ""),
                fontSize = CivTokens.Typography.FontSize.lg,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.semantics { this.heading() },
            )
        }
        if (hint.isNotEmpty()) {
            Text(hint, fontSize = CivTokens.Typography.FontSize.sm, color = hintColor)
        }
        if (error.isNotEmpty()) {
            Text(
                error,
                fontSize = CivTokens.Typography.FontSize.sm,
                color = errorColor,
                modifier = Modifier.semantics { contentDescription = error },
            )
        }

        Column(verticalArrangement = Arrangement.spacedBy(CivTokens.Spacing._3)) {
            Text("Their name", fontWeight = FontWeight.Medium)
            TextField(
                value = value.first,
                onValueChange = { onChange(value.copy(first = it)) },
                label = { Text("First name") },
                enabled = !disabled,
                readOnly = readonly,
                modifier = Modifier.fillMaxWidth(),
            )
            TextField(
                value = value.middle,
                onValueChange = { onChange(value.copy(middle = it)) },
                label = { Text("Middle name") },
                enabled = !disabled,
                readOnly = readonly,
                modifier = Modifier.fillMaxWidth(),
            )
            TextField(
                value = value.last,
                onValueChange = { onChange(value.copy(last = it)) },
                label = { Text("Last name") },
                enabled = !disabled,
                readOnly = readonly,
                modifier = Modifier.fillMaxWidth(),
            )
            if (nameError.isNotEmpty()) {
                Text(nameError, fontSize = CivTokens.Typography.FontSize.sm, color = errorColor)
            }
        }

        Column(verticalArrangement = Arrangement.spacedBy(CivTokens.Spacing._2)) {
            Text("Date of birth")
            TextField(
                value = value.dateOfBirth,
                onValueChange = { onChange(value.copy(dateOfBirth = it)) },
                placeholder = { Text("YYYY-MM-DD") },
                enabled = !disabled,
                readOnly = readonly,
                modifier = Modifier.fillMaxWidth(),
            )
            if (dateOfBirthError.isNotEmpty()) {
                Text(dateOfBirthError, fontSize = CivTokens.Typography.FontSize.sm, color = errorColor)
            }
        }

        Column(verticalArrangement = Arrangement.spacedBy(CivTokens.Spacing._2)) {
            Text("Date of death" + (if (required) " (required)" else ""))
            TextField(
                value = value.dateOfDeath,
                onValueChange = { onChange(value.copy(dateOfDeath = it)) },
                placeholder = { Text("YYYY-MM-DD") },
                enabled = !disabled,
                readOnly = readonly,
                modifier = Modifier.fillMaxWidth(),
            )
            if (dateOfDeathError.isNotEmpty()) {
                Text(dateOfDeathError, fontSize = CivTokens.Typography.FontSize.sm, color = errorColor)
            }
        }

        if (!hideRelationship) {
            Column(verticalArrangement = Arrangement.spacedBy(CivTokens.Spacing._2)) {
                Text("Their relationship to you")
                ExposedDropdownMenuBox(
                    expanded = relationshipExpanded,
                    onExpandedChange = { if (!disabled) relationshipExpanded = it },
                ) {
                    TextField(
                        value = relationshipLabel,
                        onValueChange = {},
                        readOnly = true,
                        enabled = !disabled,
                        trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = relationshipExpanded) },
                        modifier = Modifier.fillMaxWidth().menuAnchor(),
                    )
                    androidx.compose.material3.ExposedDropdownMenu(
                        expanded = relationshipExpanded,
                        onDismissRequest = { relationshipExpanded = false },
                    ) {
                        RELATIONSHIP_OPTIONS.forEach { (rValue, rLabel) ->
                            DropdownMenuItem(
                                text = { Text(rLabel) },
                                onClick = {
                                    onChange(value.copy(relationship = rValue))
                                    relationshipExpanded = false
                                },
                            )
                        }
                    }
                }
                if (relationshipError.isNotEmpty()) {
                    Text(relationshipError, fontSize = CivTokens.Typography.FontSize.sm, color = errorColor)
                }
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun CivDeceasedPersonPreview() {
    var person by remember { mutableStateOf(DeceasedPersonValue()) }
    CivDeceasedPerson(
        value = person,
        onChange = { person = it },
        legend = "About your spouse",
        required = true,
        modifier = Modifier.padding(16.dp),
    )
}

File: packages/ios/Sources/CivUI/CivDeceasedPerson.swift
// CivUI — CivDeceasedPerson for SwiftUI
// Compound field for collecting information about a person who has died.
// Used on VA burial (21P-530), SSA survivor benefit, and probate forms.
// Plain-language labels — "the person who died", not "decedent".

import SwiftUI

/// Structured value matching the web `DeceasedPersonValue` interface.
public struct DeceasedPersonValue: Equatable {
    public var first: String
    public var middle: String
    public var last: String
    public var suffix: String
    public var dateOfBirth: String
    public var dateOfDeath: String
    public var relationship: String

    public init(
        first: String = "",
        middle: String = "",
        last: String = "",
        suffix: String = "",
        dateOfBirth: String = "",
        dateOfDeath: String = "",
        relationship: String = ""
    ) {
        self.first = first
        self.middle = middle
        self.last = last
        self.suffix = suffix
        self.dateOfBirth = dateOfBirth
        self.dateOfDeath = dateOfDeath
        self.relationship = relationship
    }

    public var isEmpty: Bool {
        first.isEmpty && last.isEmpty && dateOfBirth.isEmpty && dateOfDeath.isEmpty
    }
}

private let RELATIONSHIP_OPTIONS: [(value: String, label: String)] = [
    ("", "- Select -"),
    ("spouse", "Spouse"),
    ("parent", "Parent"),
    ("child", "Child"),
    ("sibling", "Sibling"),
    ("other", "Other"),
]

public struct CivDeceasedPerson: View {
    @Binding public var value: DeceasedPersonValue

    public var legend: String
    public var name: String
    public var hint: String
    public var error: String
    public var nameError: String
    public var dateOfBirthError: String
    public var dateOfDeathError: String
    public var relationshipError: String
    public var hideRelationship: Bool
    public var required: Bool
    public var disabled: Bool
    public var readonly: Bool

    public var onChange: ((DeceasedPersonValue) -> Void)?

    @Environment(\.colorScheme) private var colorScheme

    public init(
        value: Binding<DeceasedPersonValue>,
        legend: String = "About the person who died",
        name: String = "deceased",
        hint: String = "",
        error: String = "",
        nameError: String = "",
        dateOfBirthError: String = "",
        dateOfDeathError: String = "",
        relationshipError: String = "",
        hideRelationship: Bool = false,
        required: Bool = false,
        disabled: Bool = false,
        readonly: Bool = false,
        onChange: ((DeceasedPersonValue) -> Void)? = nil
    ) {
        self._value = value
        self.legend = legend
        self.name = name
        self.hint = hint
        self.error = error
        self.nameError = nameError
        self.dateOfBirthError = dateOfBirthError
        self.dateOfDeathError = dateOfDeathError
        self.relationshipError = relationshipError
        self.hideRelationship = hideRelationship
        self.required = required
        self.disabled = disabled
        self.readonly = readonly
        self.onChange = onChange
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: CivTokens.Spacing._4) {
            if !legend.isEmpty {
                Text(legend + (required ? " (required)" : ""))
                    .font(.system(
                        size: CivTokens.Typography.FontSize.lg,
                        weight: CivTokens.Typography.FontWeight.semibold
                    ))
                    .accessibilityAddTraits(.isHeader)
            }
            if !hint.isEmpty {
                Text(hint)
                    .font(.system(size: CivTokens.Typography.FontSize.sm))
                    .foregroundColor(
                        colorScheme == .dark
                            ? CivTokens.DarkColors.Base.dark
                            : CivTokens.Colors.Base.dark
                    )
            }
            if !error.isEmpty {
                Text(error)
                    .font(.system(size: CivTokens.Typography.FontSize.sm))
                    .foregroundColor(
                        colorScheme == .dark
                            ? CivTokens.DarkColors.Error.default_
                            : CivTokens.Colors.Error.default_
                    )
                    .accessibilityAddTraits(.isStaticText)
            }

            VStack(alignment: .leading, spacing: CivTokens.Spacing._3) {
                Text("Their name")
                    .font(.system(size: CivTokens.Typography.FontSize.base, weight: .medium))
                TextField("First name", text: $value.first)
                    .textContentType(.givenName)
                    .disabled(disabled)
                TextField("Middle name", text: $value.middle)
                    .textContentType(.middleName)
                    .disabled(disabled)
                TextField("Last name", text: $value.last)
                    .textContentType(.familyName)
                    .disabled(disabled)
                if !nameError.isEmpty {
                    Text(nameError)
                        .font(.system(size: CivTokens.Typography.FontSize.sm))
                        .foregroundColor(
                            colorScheme == .dark
                                ? CivTokens.DarkColors.Error.default_
                                : CivTokens.Colors.Error.default_
                        )
                }
            }

            VStack(alignment: .leading, spacing: CivTokens.Spacing._2) {
                Text("Date of birth")
                TextField("YYYY-MM-DD", text: $value.dateOfBirth)
                    .disabled(disabled)
                if !dateOfBirthError.isEmpty {
                    Text(dateOfBirthError)
                        .font(.system(size: CivTokens.Typography.FontSize.sm))
                        .foregroundColor(
                            colorScheme == .dark
                                ? CivTokens.DarkColors.Error.default_
                                : CivTokens.Colors.Error.default_
                        )
                }
            }

            VStack(alignment: .leading, spacing: CivTokens.Spacing._2) {
                Text("Date of death" + (required ? " (required)" : ""))
                TextField("YYYY-MM-DD", text: $value.dateOfDeath)
                    .disabled(disabled)
                if !dateOfDeathError.isEmpty {
                    Text(dateOfDeathError)
                        .font(.system(size: CivTokens.Typography.FontSize.sm))
                        .foregroundColor(
                            colorScheme == .dark
                                ? CivTokens.DarkColors.Error.default_
                                : CivTokens.Colors.Error.default_
                        )
                }
            }

            if !hideRelationship {
                VStack(alignment: .leading, spacing: CivTokens.Spacing._2) {
                    Text("Their relationship to you")
                    Picker("Relationship", selection: $value.relationship) {
                        ForEach(RELATIONSHIP_OPTIONS, id: \.value) { option in
                            Text(option.label).tag(option.value)
                        }
                    }
                    .pickerStyle(.menu)
                    .disabled(disabled)
                    if !relationshipError.isEmpty {
                        Text(relationshipError)
                            .font(.system(size: CivTokens.Typography.FontSize.sm))
                            .foregroundColor(
                                colorScheme == .dark
                                    ? CivTokens.DarkColors.Error.default_
                                    : CivTokens.Colors.Error.default_
                            )
                    }
                }
            }
        }
        .onChange(of: value) { newValue in
            onChange?(newValue)
        }
    }
}

#if DEBUG
struct CivDeceasedPerson_Previews: PreviewProvider {
    struct Wrapper: View {
        @State var person = DeceasedPersonValue()
        var body: some View {
            CivDeceasedPerson(value: $person, legend: "About your spouse", required: true)
                .padding()
        }
    }
    static var previews: some View { Wrapper() }
}
#endif
